File: 60058_괄호변환/Programmers60058.java
public class Programmers60058 {
    public static void main(String args[]) {

    }
}

class Solution {
    private String[] splitBalanced(String w) {
        int leftCnt = 0;
        int rightCnt = 0;
        int splitIdx = 0;
        for (int i = 0; i < w.length(); i++) {
            if (w.charAt(i) == '(') {
                leftCnt++;
            } else {
                rightCnt++;
            }
            if (leftCnt == rightCnt) {
                splitIdx = i;
                break;
            }
        }
        return new String[]{w.substring(0, splitIdx + 1), w.substring(splitIdx + 1)};
    }

    private boolean isRightString(String u) {
        int leftCnt = 0;
        int rightCnt = 0;
        for (int i = 0; i < u.length(); i++) {
            if (u.charAt(i) == '(') leftCnt++;
            else rightCnt++;
            if (rightCnt > leftCnt) return false;
        }
        return leftCnt == rightCnt;
    }

    private String flip(String u) {
        StringBuilder flipU = new StringBuilder();
        for (int i = 0; i < u.length(); i++) {
            flipU.append(u.charAt(i) == '(' ? ')' : '(');
        }
        return new String(flipU);
    }

    private String change(String p) {
        if (p.isEmpty())
            return p;
        String[] uv = splitBalanced(p);
        if (isRightString(uv[0])) {
            return uv[0] + change(uv[1]);
        } else {
            String v = "(" + change(uv[1]) + ")";
            String u = flip(uv[0].substring(1, uv[0].length() - 1));
            return (v + u);
        }
    }

    public String solution(String p) {
        String answer = change(p);
        return answer;
    }
}

File: Programmers/60058_괄호변환/Programmers60058.py
def checkRightStr(u):
    leftCnt = 0
    rightCnt = 0
    for c in u:
        if c == '(':
            leftCnt += 1
        else:
            rightCnt += 1
        if leftCnt < rightCnt:
            return False
    return leftCnt == rightCnt

def flip(x):
    flipHash = {"(":")", ")":"("}
    return flipHash[x]

def change(w):
    if w == "": return w
    u, v = splitBalenced(w)
    if checkRightStr(u):
        return u + change(v)
    else:
        u = "".join([flip(x) for x in u[1:-1]])
        u = "(" + change(v) + ")" + u
        return u

    
def splitBalenced(w):
    leftCnt = 0
    rightCnt = 0
    for i, c in enumerate(w):
        if c == '(':
            leftCnt += 1
        else:
            rightCnt += 1
        if leftCnt == rightCnt:
            return (w[0:i+1], w[i+1:])

def solution(p):
    answer = change(p)
    return answer
